### ISSUE TITLE
Code Style and Naming

### ISSUE THREAD

**Issue by robintibor** [braindecode/braindecode#59]
When we have the first integrated version skorch-api+dataset, we can make a pass over the code to enforce some more consistency and see what style we prefer (beyond just pep8).

Let's collect open questions for now:
* to indicate a variable that specifies an index/counter, e.g., the index of a trial, should we prefix with `i`(=`i_trial`) or with `idx`=`idx_trial`

**Comment by agramfort** [comment]
my only suggestion would be to use n_xxx to denote cardinalities (n_samples, n_features, n_epochs etc.)

**Comment by hubertjb** [comment]
I created a wiki page to list these naming conventions: https://github.com/braindecode/braindecode/wiki/Variable-naming-conventions

I would vote for `i_xxx` instead of `idx_xxx`, and to use `n_xxx` as suggested by @agramfort.

**Comment by robintibor** [comment]
In the same pass as code style, we can also do naming stuff.
There we have to decide about supercrop-language, if we want to keep it and where etc.

**Comment by robintibor** [comment]
Another Point:
* Expression -> Lambda?

**Comment by agramfort** [comment]
Expression is ok for me.

>

**Comment by hubertjb** [comment]
Working on #65 I've thought a bit more about naming conventions for supercrops vs. windows.

So we currently have the following concepts:
- Trial
- Supercrop
- Crop

Here are some thoughts:
- "Trials" are specific to recordings with some kind of paradigm or stimulus presentation (e.g., a BCI experiment). Sleep recordings, or clinical recordings (e.g., TUH) don't usually have trials, but instead have annotated regions, such as sleep stages, or arousal/seizure/etc. annotations. To avoid confusion (e.g., a user might wonder what a trial means in the context of sleep data?), I think what is currently called "trials" could instead be renamed "annotations". It also maps nicely to the concept of annotations (and annotations objects) in MNE.
- I think we agree that "supercrops" and "windows" mean the same thing (and both are used so far in the code). I would personally pick "windows" over "supercrops", as it's a more general term that anyone with basic knowledge of signal processing concepts should understand. I see "supercrops" as  more of an implementation-specific term. Also, I think we agree that the other alternative, "epochs", is to be avoided given its meaning in DL.
- As for "crops", I believe the term currently only appears in samplers and models that explicitly require this concept. It makes sense to keep the work "crops" in this case.

Let me know what you guys think @robintibor @gemeinl @sbbrandt . When we've decided on a convention to follow I can open a PR to standardize functions and variable names.

**Comment by agramfort** [comment]
one word that pops to mind is also "segment" that point more directly to
data / signals while annotations refers to the target variable.

>

**Comment by robintibor** [comment]
I think one issues is we have multiple things which refer to timewindows and differ in the meaning of the timewindows:
* Trials, as @hubertjb  said, tied to a specific stimulus presentation/task
* Supercrops, tied to deep learning training loop
* Crops, tied to receptive field of deep network

How about we make this more explicit like:
Trial -> Trialwindow
Supercrop -> CropsWindow (or TrainingWindow? Maybe much more explicit?)

**Comment by hubertjb** [comment]
I was actually thinking we should use terminology that is general enough to apply to the main use cases of the package. As long as we explicitly define these terms in the documentation, we should be good.

> How about we make this more explicit like:
Trial -> TrialWindow

But then we would still have the word "trial" in there. I'm thinking specifically about this function: https://github.com/braindecode/braindecode/blob/master/braindecode/datautil/windowers.py#L18
Should we keep the word "trial"? I think "segment" as @agramfort suggested would apply to sleep, seizure annotations, etc. as well as BCI trials.

> Supercrop -> CropsWindow (or TrainingWindow? Maybe much more explicit? PredictWindow?)

Why not just "windows"?

I suggest we include a little drawing in the documentation, along with some defintions. For example, I was thinking we could use the following terminology:
![terminology](https://user-images.githubusercontent.com/6500858/74172944-7186f580-4c31-11ea-9cf2-3e2d97d82e08.jpg)

* **Segment**: A chunk of a recording that has been annotated, e.g., a trial in a stimulus-driven paradigm such as a BCI experiment, a sleep stage annotation, a seizure event, etc.
* **Window**: A group of consecutive EEG samples extracted for further analysis, typically between 0.5 and 30 s, and that are used as input to a neural network. Windows are sometimes called "supercrops" in the deep learning literature [refs] or "epochs" in the EEG processing literature [refs].
* **Crop**: A subdivision of a window, used in convolutional architectures to improve computational efficiency at training time. [refs]

We initially tried to define everything in the DL-EEG review, although some changes are required now:
![image](https://user-images.githubusercontent.com/6500858/74173680-d1ca6700-4c32-11ea-8e32-fb68d3cc6c3b.png)
![image](https://user-images.githubusercontent.com/6500858/74173716-e149b000-4c32-11ea-9c53-db46e06ea9cf.png)

**Comment by robintibor** [comment]
So my view atm:
* Trial
* ComputeWindow
* ReceptiveField (possibly only needed in very few places)

**Comment by robintibor** [comment]
Thinking about it if we use ComputeCrop and this is only time we use crop, inside the code we could shorten to crop?

**Comment by sliwy** [comment]
@robintibor what do you mean by "computational aspects", can you explain better? What I understand is that compute window is a single input to the network? I wonder if it change interpretation of the variable name?

**Comment by robintibor** [comment]
yes it is a single input to the network. the larger you make this input, the more predictions you will get -> the more are computed in parallel, saving you computational time.

**Comment by sliwy** [comment]
I feel like windows are more important for problem definition than for computations, what do you think? From my perspective, this is how we define the EEG signal fragment for which we compute predictions and  overall loss (we have predictions for all receptive_fields and average them to obtain final window prediction)? I would opt for "window" without the prefix, as atm I don't see a potential profit from adding "compute".
For the trial and receptive fields I think, they are quite clear. Trial was already introduced and it is easy to understand. Receptive field name shows that it is quite internal (the name is not so general and strictly connected to cropped decoding). Unfortunately, there is no connection in naming "cropped decoding" <-> "receptive field", so maybe we will need to change it somehow :/

**Comment by robintibor** [comment]
Unfortunately, "we have predictions for all receptive_fields and average them to obtain final window prediction" is not correct in general. For example, one use case that we had before and should continue to support is regressing a time series of a target. E.g. you have an EEG signal and a corresponding timeseries of a changing target, like position of a steering wheel the EEG user is controlling. In that case you only create windows to make computations feasible (can't input entire recording at once into the network), but the size of that window has no semantic meaning whatsoever, it is a purely computational choice. And actually for the case of trials with a single label where you cut computewindows inside, one can still view it equivalently in one of our main settings. Like you can also do it by never averaging the predictions, rather evaluating the loss for each prediction and then averaging the losses. this yields identical gradients in some settings (log softmax outputs / cross entropy loss).

**Comment by robintibor** [comment]
See https://onlinelibrary.wiley.com/doi/full/10.1002/hbm.23730 -> ConvNet Training-> Cropped Training for original motivation

**Comment by robintibor** [comment]
based on all of this how about:
* Trial
* (Compute) Window -> we use compute_window at some public function calls to give a strong hint at the beginning, internally can use shorter "window"
* Crop, stays as crop to stay consistent with paper, mostly used in internal code

**Comment by robintibor** [comment]
So we could still go through some namings this sprint if we want, but in a backward-compatible way via depracation. like the `n_`  logic for cardinalities etc... feel unsure about whether better to change or just keep as is... @hubertjb @sliwy @agramfort @gemeinl

**Comment by robintibor** [comment]
See https://github.com/braindecode/braindecode/issues/28 also
